pkg/capabilities/capabilities.go

```diff
@@ -1,88 +1,295 @@
+	c.set(Required, values...) // populate ring1 (Required)
 }
 
-func checkRequired(caps set, reqCaps []cap.Value) error {
-	var missingCaps []cap.Value
-	for _, c := range reqCaps {
-		exist, err := caps.GetFlag(cap.Effective, c)
-		if err != nil {
-			return err
+func (c *Capabilities) RequireByString(values ...string) error {
+	var capsToAdd []cap.Value
+
+	for _, cap := range values {
+		found := false
+		for k := range c.all {
+			if k.String() == cap {
+				capsToAdd = append(capsToAdd, k)
+				found = true
+			}
 		}
-		if !exist {
-			missingCaps = append(missingCaps, c)
+		if !found {
+			return couldNotFindCapability(cap)
 		}
 	}
-	if len(missingCaps) > 0 {
-		return &MissingCapabilitiesError{missingCaps}
+
+	for _, cap := range capsToAdd {
+		logger.Debug("capability required by user", "pkg", pkgName, "cap", cap.String())
 	}
+
+	c.Require(capsToAdd...)
+
 	return nil
 }
 
-// DropUnrequired drops all capabilities not required by user from Effective and Permitted set, and all capabilities
-// from the Inheritance set.
-// DropUnrequired requires that all required capabilities are already set or available in permitted set.
-// The function also tries to drop the capabilities bounding set, but it won't work if CAP_SETPCAP is not available.
-func DropUnrequired(reqCaps []cap.Value) error {
-	selfCaps, err := getCurrentProcessCapabilities()
-	if err != nil {
-		return err
+// Unrequire is only called when command line "capabilities drop=X" is given.
+// It works by removing, from the required ring, the capabilities given by the
+// user. This way, when tracee shifts to ring1 (Required), that capability won't
+// be Effective.
+func (c *Capabilities) Unrequire(values ...cap.Value) {
+	if c.bypass {
+		return
 	}
-	return dropUnrequired(selfCaps, reqCaps)
+	c.unset(Required, values...) // unpopulate ring1 (Required)
 }
 
-func dropUnrequired(selfCaps set, reqCaps []cap.Value) error {
-	// Dropping the bounding set is a best effort, so we ignore any error resulted from doing it.
-	cap.DropBound(GetAllCapabilities()...)
-	err := selfCaps.Clear()
-	if err != nil {
-		return err
+func (c *Capabilities) UnrequireByString(values ...string) error {
+	var capsToDrop []cap.Value
+
+	for _, cap := range values {
+		found := false
+		for k := range c.all {
+			if k.String() == cap {
+				capsToDrop = append(capsToDrop, k)
+				found = true
+			}
+		}
+		if !found {
+			return couldNotFindCapability(cap)
+		}
 	}
-	err = selfCaps.SetFlag(cap.Effective, true, reqCaps...)
-	if err != nil {
-		return err
+
+	for _, cap := range capsToDrop {
+		logger.Debug("capability UN-required by user", "pkg", pkgName, "cap", cap.String())
 	}
-	err = selfCaps.SetFlag(cap.Permitted, true, reqCaps...)
+
+	c.Unrequire(capsToDrop...)
+
+	return nil
+}
+
+// internal
+
+func (c *Capabilities) getProc() {
+	c.have = cap.GetProc()
+}
+
+func (c *Capabilities) setProc() {
+	err := c.have.SetProc()
 	if err != nil {
-		return err
+		panic(err)
```

Comment:
Why panic here?

Reply:
We can handle the errors up to last scope and then exit with a Clean() but I thought... no capabilities might get me into a situation where I can't clean =D.

Reply:
We can always panic on the failure outside (in `tracee.Init()`), but there's also merit to the idea of not letting any attack surface at all here. @yanivagman WDYT?

Reply:
That's a good question.
In the future, this won't necessarily be called only from Init, right?
Anyway, I don't like much the option of panicing Tracee. It can happen that something will fail due to some bug, and then we crash Tracee although we could have possibly continue the execution with the existing capabilties.
I'm in favor of logging an error here.

Reply:
Alright! Lets error instead!